File: src/generate/compose.ts
import { dirname, join } from "node:path";
import { expandAll } from "../core/interpolate.ts";
import type { PlanIssue } from "../core/plan.ts";
import type { Registry } from "../registry/types.ts";
import type { VaultSession } from "../vault/provider.ts";
import { globalsFor } from "./generate.ts";
import { disclaimerHeader } from "./ownership.ts";

export interface MarkerRegion {
  consumer: string;
  start: number; // index of the opening `# <menv:consumer>` line
  end: number; // index of the closing `# </menv>` line
  indent: string; // leading whitespace of the opening marker
}

const OPEN_RE = /^(\s*)#\s*<menv:([a-z0-9][a-z0-9._-]*)>\s*$/;
const CLOSE_RE = /^\s*#\s*<\/menv>\s*$/;

// Markers are hand-authored by the user; menv only discovers them and rewrites
// the lines between each pair. Structural errors are reported, never fixed.
export function findMarkerRegions(content: string): { regions: MarkerRegion[]; errors: string[] } {
  const lines = content.split("\n");
  const regions: MarkerRegion[] = [];
  const errors: string[] = [];
  let open: { consumer: string; start: number; indent: string } | null = null;
  for (let i = 0; i < lines.length; i++) {
    const line = lines[i] ?? "";
    const openM = line.match(OPEN_RE);
    if (openM) {
      if (open !== null) errors.push(`nested menv marker at line ${i + 1}`);
      open = { consumer: openM[2] as string, start: i, indent: openM[1] as string };
      continue;
    }
    if (CLOSE_RE.test(line)) {
      if (open === null) {
        errors.push(`unmatched </menv> at line ${i + 1}`);
        continue;
      }
      regions.push({ consumer: open.consumer, start: open.start, end: i, indent: open.indent });
      open = null;
    }
  }
  if (open !== null) errors.push(`unclosed <menv:${open.consumer}> marker`);
  return { regions, errors };
}

// Replaces each region's body (the lines strictly between its markers) with the
// supplied fill lines, keyed by the region's start index. Lines outside every
// region — including the marker lines themselves — are preserved verbatim.
export function spliceRegions(content: string, regions: MarkerRegion[], fillByStart: Map<number, string[]>): string {
  const lines = content.split("\n");
  const out: string[] = [];
  let i = 0;
  const byStart = new Map(regions.map((r) => [r.start, r]));
  while (i < lines.length) {
    const region = byStart.get(i);
    if (region !== undefined) {
      out.push(lines[i] as string); // opening marker
      out.push(...(fillByStart.get(region.start) ?? []));
      out.push(lines[region.end] as string); // closing marker
      i = region.end + 1;
      continue;
    }
    out.push(lines[i] as string);
    i += 1;
  }
  return out.join("\n");
}

// The interpolation key for a (consumer, variable): consumer-prefixed so two
// services sharing one .env.compose never collide.
export function composeKey(consumer: string, name: string): string {
  return `${consumer.toUpperCase().replace(/[^A-Z0-9]/g, "_")}_${name}`;
}

interface ComposeValue {
  key: string;
  value: string;
  disabled: boolean;
}

export interface ComposePreview {
  writes: { path: string; content: string }[];
  errors: PlanIssue[];
  warnings: PlanIssue[];
}

// One compose pass over every registered file for the selected vault. Each
// file's marker regions are filled and a sibling .env.compose (the union of
// every region's values, disabled commented) is rendered.
export async function previewCompose(
  root: string,
  registry: Registry,
  opts: { vault?: string },
  sessions: ReadonlyMap<string, VaultSession>,
): Promise<ComposePreview> {
  const vault = opts.vault ?? registry.defaults.vault;
  const preview: ComposePreview = { writes: [], errors: [], warnings: [] };
  const session = sessions.get(vault);
  const globals = globalsFor(registry, vault);
  // Group compose files by directory so one .env.compose serves all files there.
  const valuesByDir = new Map<string, Map<string, ComposeValue>>();
  const splicedWrites: { path: string; content: string }[] = [];

  for (const file of registry.compose.files) {
    const abs = Bun.file(join(root, file));
    if (!(await abs.exists())) {
      preview.errors.push({ code: "MISSING_COMPOSE_FILE", message: `registered compose file not found: ${file}` });
      continue;
    }
    const content = await abs.text();
    const { regions, errors } = findMarkerRegions(content);
    for (const e of errors) preview.errors.push({ code: "COMPOSE_MARKER", message: `${file}: ${e}` });
    if (errors.length > 0) continue;
    if (regions.length === 0) {
      preview.warnings.push({ code: "COMPOSE_NO_MARKERS", message: `${file}: bound but has no menv markers` });
    }
    const dir = dirname(file) === "." ? "" : dirname(file);
    const dirValues = valuesByDir.get(dir) ?? new Map<string, ComposeValue>();
    valuesByDir.set(dir, dirValues);
    const fillByStart = new Map<number, string[]>();
    let fileFailed = false;
    for (const region of regions) {
      if (registry.consumers[region.consumer] === undefined) {
        preview.errors.push({
          code: "COMPOSE_UNKNOWN_CONSUMER",
          message: `${file}: marker names unknown consumer "${region.consumer}"`,
        });
        fileFailed = true;
        continue;
      }
      if (session === undefined) {
        preview.warnings.push({ code: "UNVERIFIED_VAULT", message: `vault "${vault}" could not be opened for compose` });
        fileFailed = true;
        continue;
      }
      const raw = new Map<string, string>();
      const meta: { name: string; disabled: boolean }[] = [];
      const names = Object.keys(registry.variables)
        .filter((n) => registry.variables[n]?.vaultMapping[vault]?.[region.consumer] !== undefined)
        .sort();
      for (const name of names) {
        const entry = registry.variables[name]?.vaultMapping[vault]?.[region.consumer];
        if (entry === undefined) continue;
        raw.set(name, (await session.get(entry.key)) ?? "");
        meta.push({ name, disabled: entry.disabled === true });
      }
      const expanded = expandAll({ values: raw, globals });
      const fill: string[] = [];
      for (const { name, disabled } of meta) {
        const key = composeKey(region.consumer, name);
        fill.push(`${region.indent}- ${name}=\${${key}}`);
        dirValues.set(key, { key, value: expanded.get(name) ?? "", disabled });
      }
      fillByStart.set(region.start, fill);
    }
    if (fileFailed) continue;
    splicedWrites.push({ path: file, content: spliceRegions(content, regions, fillByStart) });
  }

  if (preview.errors.length > 0) return { ...preview, writes: [] };
  preview.writes.push(...splicedWrites);
  for (const [dir, values] of valuesByDir) {
    const header = disclaimerHeader({ vault });
    const lines = [...values.values()]
      .sort((a, b) => a.key.localeCompare(b.key))
      .map((v) => (v.disabled ? `# ${v.key}=${v.value}` : `${v.key}=${v.value}`));
    preview.writes.push({ path: join(dir, ".env.compose"), content: `${header}${lines.join("\n")}\n` });
  }
  return preview;
}

File: go/internal/generate/compose.go
package generate

import (
	"fmt"
	"path/filepath"
	"regexp"
	"sort"
	"strings"

	"github.com/nikrabaev/menv/go/internal/core"
	"github.com/nikrabaev/menv/go/internal/registry"
)

var (
	openRE  = regexp.MustCompile(`^(\s*)#\s*<menv:([a-z0-9][a-z0-9._-]*)>\s*$`)
	closeRE = regexp.MustCompile(`^\s*#\s*</menv>\s*$`)
)

// MarkerRegion describes one pair of menv compose markers within a file.
type MarkerRegion struct {
	Consumer string
	Start    int    // index of the opening `# <menv:consumer>` line
	End      int    // index of the closing `# </menv>` line
	Indent   string // leading whitespace of the opening marker
}

// FindMarkerRegions parses compose markers from file content.
func FindMarkerRegions(content string) (regions []MarkerRegion, errors []string) {
	lines := strings.Split(content, "\n")
	type openState struct {
		consumer string
		start    int
		indent   string
	}
	var open *openState
	for i, line := range lines {
		if m := openRE.FindStringSubmatch(line); m != nil {
			if open != nil {
				errors = append(errors, fmt.Sprintf("nested menv marker at line %d", i+1))
			}
			open = &openState{consumer: m[2], start: i, indent: m[1]}
			continue
		}
		if closeRE.MatchString(line) {
			if open == nil {
				errors = append(errors, fmt.Sprintf("unmatched </menv> at line %d", i+1))
				continue
			}
			regions = append(regions, MarkerRegion{
				Consumer: open.consumer,
				Start:    open.start,
				End:      i,
				Indent:   open.indent,
			})
			open = nil
		}
	}
	if open != nil {
		errors = append(errors, fmt.Sprintf("unclosed <menv:%s> marker", open.consumer))
	}
	return regions, errors
}

// SpliceRegions replaces the body between each marker pair with the supplied
// fill lines; marker lines themselves and all other content are preserved.
func SpliceRegions(content string, regions []MarkerRegion, fillByStart map[int][]string) string {
	lines := strings.Split(content, "\n")
	byStart := make(map[int]MarkerRegion, len(regions))
	for _, r := range regions {
		byStart[r.Start] = r
	}
	var out []string
	i := 0
	for i < len(lines) {
		if r, ok := byStart[i]; ok {
			out = append(out, lines[i]) // opening marker
			out = append(out, fillByStart[r.Start]...)
			out = append(out, lines[r.End]) // closing marker
			i = r.End + 1
			continue
		}
		out = append(out, lines[i])
		i++
	}
	return strings.Join(out, "\n")
}

// ComposeKey returns the interpolation key for a (consumer, variable) pair.
func ComposeKey(consumer, name string) string {
	upper := strings.ToUpper(consumer)
	var b strings.Builder
	for _, ch := range upper {
		if (ch >= 'A' && ch <= 'Z') || (ch >= '0' && ch <= '9') {
			b.WriteRune(ch)
		} else {
			b.WriteRune('_')
		}
	}
	b.WriteRune('_')
	b.WriteString(name)
	return b.String()
}

// ComposePreview is the result of one compose generation pass.
type ComposePreview struct {
	Writes   []FileWrite
	Errors   []core.PlanIssue
	Warnings []core.PlanIssue
}

type composeValue struct {
	key      string
	value    string
	disabled bool
}

// PreviewCompose processes all registered compose files, filling marker regions
// and computing .env.compose content for each directory.
func PreviewCompose(
	root string,
	reg registry.Registry,
	opts GenerateOpts,
	sessions map[string]core.VaultSession,
) (ComposePreview, error) {
	vaultName := coalesce(opts.Vault, reg.Defaults.Vault)
	preview := ComposePreview{}
	sess := sessions[vaultName]
	globals := GlobalsFor(reg, vaultName)

	type dirEntry struct {
		values map[string]composeValue
	}
	valuesByDir := map[string]*dirEntry{}
	succeededDirs := map[string]bool{}
	var splicedWrites []FileWrite

	for _, file := range reg.Compose.Files {
		data, err := readFile(filepath.Join(root, file))
		if err != nil {
			preview.Errors = append(preview.Errors, core.PlanIssue{
				Code:    "MISSING_COMPOSE_FILE",
				Message: fmt.Sprintf("registered compose file not found: %s", file),
			})
			continue
		}
		content := string(data)
		regions, errs := FindMarkerRegions(content)
		for _, e := range errs {
			preview.Errors = append(preview.Errors, core.PlanIssue{Code: "COMPOSE_MARKER", Message: file + ": " + e})
		}
		if len(errs) > 0 {
			continue
		}
		if len(regions) == 0 {
			preview.Warnings = append(preview.Warnings, core.PlanIssue{
				Code:    "COMPOSE_NO_MARKERS",
				Message: file + ": bound but has no menv markers",
			})
		}
		dir := filepath.Dir(file)
		if dir == "." {
			dir = ""
		}
		if _, ok := valuesByDir[dir]; !ok {
			valuesByDir[dir] = &dirEntry{values: map[string]composeValue{}}
		}
		de := valuesByDir[dir]
		fillByStart := map[int][]string{}
		fileFailed := false

		for _, region := range regions {
			if _, ok := reg.Consumers[region.Consumer]; !ok {
				preview.Errors = append(preview.Errors, core.PlanIssue{
					Code:    "COMPOSE_UNKNOWN_CONSUMER",
					Message: fmt.Sprintf("%s: marker names unknown consumer %q", file, region.Consumer),
				})
				fileFailed = true
				continue
			}
			if sess == nil {
				preview.Warnings = append(preview.Warnings, core.PlanIssue{
					Code:    "UNVERIFIED_VAULT",
					Message: fmt.Sprintf("vault %q could not be opened for compose", vaultName),
				})
				fileFailed = true
				continue
			}

			rawVals := map[string]string{}
			type varMeta struct {
				name     string
				disabled bool
			}
			var varMetas []varMeta
			// Collect wired variable names in sorted order.
			varNames := make([]string, 0)
			for name, def := range reg.Variables {
				if _, ok := def.VaultMapping[vaultName][region.Consumer]; ok {
					varNames = append(varNames, name)
				}
			}
			sort.Strings(varNames)
			for _, name := range varNames {
				entry := reg.Variables[name].VaultMapping[vaultName][region.Consumer]
				val, _, err := sess.Get(entry.Key)
				if err != nil {
					return ComposePreview{}, err
				}
				rawVals[name] = val
				varMetas = append(varMetas, varMeta{name: name, disabled: entry.Disabled})
			}
			expanded, err := core.ExpandAll(core.ExpandInput{Values: rawVals, Globals: globals})
			if err != nil {
				return ComposePreview{}, err
			}
			var fill []string
			for _, vm := range varMetas {
				key := ComposeKey(region.Consumer, vm.name)
				fill = append(fill, fmt.Sprintf("%s- %s=${%s}", region.Indent, vm.name, key))
				de.values[key] = composeValue{key: key, value: expanded[vm.name], disabled: vm.disabled}
			}
			fillByStart[region.Start] = fill
		}
		if fileFailed {
			continue
		}
		succeededDirs[dir] = true
		splicedWrites = append(splicedWrites, FileWrite{Path: file, Content: SpliceRegions(content, regions, fillByStart)})
	}

	if len(preview.Errors) > 0 {
		return ComposePreview{Errors: preview.Errors, Warnings: preview.Warnings}, nil
	}
	preview.Writes = append(preview.Writes, splicedWrites...)

	for dir, de := range valuesByDir {
		if !succeededDirs[dir] {
			continue
		}
		header := DisclaimerHeader(HeaderMeta{Vault: vaultName})
		keys := make([]string, 0, len(de.values))
		for k := range de.values {
			keys = append(keys, k)
		}
		sort.Strings(keys)
		var lines []string
		for _, k := range keys {
			cv := de.values[k]
			if cv.disabled {
				lines = append(lines, "# "+k+"="+cv.value)
			} else {
				lines = append(lines, k+"="+cv.value)
			}
		}
		composePath := filepath.Join(dir, ".env.compose")
		if dir == "" {
			composePath = ".env.compose"
		}
		content := header + strings.Join(lines, "\n") + "\n"
		preview.Writes = append(preview.Writes, FileWrite{Path: composePath, Content: content})
	}
	return preview, nil
}
